### PR TITLE
Add some JDK annotations for Signedness Checker

### PR DESCRIPTION
**Pull request by gokaco**

### CODE REVIEW

**Comment by gokaco** [comment]
The changes are working fine in my machine and ./gradlew allTests also passed then why travis is giving error here?

**Comment by gokaco** [comment]
InputStream in is getting null value and is not getting read.

**Comment by mernst** [comment]
Master was broken for a time today. :-(
I pulled master and I believe this will resolve the problem.

File: checker/src/main/java/org/checkerframework/checker/signedness/jdk.astub [hunk 1]
```diff
@@ -94,6 +98,11 @@ class Long {
     public static @Unsigned long remainderUnsigned(@Unsigned long dividend, @Unsigned long divisor);
 }
 
+class Object {
+    protected @Unsigned Object clone() throws CloneNotSupportedException;
```
Comment (mernst) [inline checker/src/main/java/org/checkerframework/checker/signedness/jdk.astub]: `@Unsigned` should not be applied to non-numeric types.

File: checker/src/main/java/org/checkerframework/checker/signedness/jdk.astub [hunk 2]
```diff
@@ -193,6 +202,18 @@ class Arrays {
     public static @PolySigned float[] copyOfRange(@PolySigned float[] original, int from, int to);
     public static @PolySigned double[] copyOfRange(@PolySigned double[] original, int from, int to);
     public static @PolySigned boolean[] copyOfRange(@PolySigned boolean[] original, int from, int to);
+    public static <T> @PolySigned T[] copyOfRange(@PolySigned T[] original, int from, int to);
```
Comment (mernst) [inline checker/src/main/java/org/checkerframework/checker/signedness/jdk.astub]: Why is this necessary?

File: checker/src/main/java/org/checkerframework/checker/signedness/jdk.astub [hunk 3]
```diff
@@ -193,6 +202,18 @@ class Arrays {
...
+
+    public static @PolySigned byte[] copyOf(@PolySigned byte[] original, int newLength);
+    public static @PolySigned short[] copyOf(@PolySigned short[] original, int newLength);
+    public static @PolySigned int[] copyOf(@PolySigned int[] original, int newLength);
+    public static @PolySigned long[] copyOf(@PolySigned long[] original, int newLength);
+    public static @PolySigned char[] copyOf(@PolySigned char[] original, int newLength);
+    public static @PolySigned float[] copyOf(@PolySigned float[] original, int newLength);
+    public static @PolySigned double[] copyOf(@PolySigned double[] original, int newLength);
+    public static @PolySigned boolean[] copyOf(@PolySigned boolean[] original, int newLength);
+    public static <T> @PolySigned T[] copyOf(@PolySigned T[] original, int newLength);
```
Comment (mernst) [inline checker/src/main/java/org/checkerframework/checker/signedness/jdk.astub]: Why is this necessary?

File: checker/src/main/java/org/checkerframework/checker/signedness/jdk.astub (hunk 1 above)
Comment (gokaco) [reply checker/src/main/java/org/checkerframework/checker/signedness/jdk.astub]: Array is an object in java and if we want to clone an integer or byte array that is unsigned then due to this change we will not get error?
for ex.-
import java.util.Arrays; 
import org.checkerframework.checker.signedness.qual.*;
class ex { 

	private static @Unsigned byte[] de(@Unsigned byte[] ci){

		@Unsigned byte b[]=ci.clone();
		return b;
	}
}
This will give error if this change is not made.

File: checker/src/main/java/org/checkerframework/checker/signedness/jdk.astub (hunk 3 above)
Comment (gokaco) [reply checker/src/main/java/org/checkerframework/checker/signedness/jdk.astub]: Yes, I think line 215 is not necessary. I just annotated all the copyOf functions. Sorry for the mistake.
And yes float double boolean cant be unsigned so their is no use of them.

File: checker/src/main/java/org/checkerframework/checker/signedness/jdk.astub (hunk 2 above)
Comment (gokaco) [reply checker/src/main/java/org/checkerframework/checker/signedness/jdk.astub]: Yes, I think line 205 is not necessary. Sorry for the mistake.

File: checker/src/main/java/org/checkerframework/checker/signedness/jdk.astub (hunk 1 above)
Comment (mernst) [reply checker/src/main/java/org/checkerframework/checker/signedness/jdk.astub]: I agree the change to the annotated JDK will fix this use case.
However, it will break others:
 * cloning a signed byte array will result in an unsigned byte array.  This introduces unsoundness.
 * cloning a string will result in an unsigned string.  This introduces illegal annotations.

It is not acceptable to fix one problem while introducing both unsoundness and illegal annotations.

Comment (mernst) [reply checker/src/main/java/org/checkerframework/checker/signedness/jdk.astub]: Can you think of a different way to achieve your goals?

Comment (gokaco) [reply checker/src/main/java/org/checkerframework/checker/signedness/jdk.astub]: Yes, If we use @PolySigned then it should work fine.

Comment (mernst) [reply checker/src/main/java/org/checkerframework/checker/signedness/jdk.astub]: Yes, that is a better choice.  Please write a few test cases to ensure that the annotations behave as you expect.  Thanks!

Comment (gokaco) [reply checker/src/main/java/org/checkerframework/checker/signedness/jdk.astub]: Yes, I checked for ex-
import java.lang.*; 
import org.checkerframework.checker.signedness.qual.*;
class ex2 { 

	private static @Unsigned byte[] de(byte[] ci){

		@Unsigned byte b[]=ci.clone();
		return b;
	}
}
This is giving error which would not occur earlier due to @Unsigned, but now due to @PolySigned it is giving the error. Rest all are working fine.

Comment (mernst) [reply checker/src/main/java/org/checkerframework/checker/signedness/jdk.astub]: Thanks for checking that.  It is usually a good idea to have both positive and negative test cases (that is, both code that typechecks and code that produces a type-checking error), and to try with multiple types (including non-arrays which was the original concern).

It may be that your changes are correct, but it's good to double-check it.
Thanks!

Comment (gokaco) [reply checker/src/main/java/org/checkerframework/checker/signedness/jdk.astub]: Sorry, but I am unable to understand "non arrays" such as? Can you give examples.
Thanks!

Comment (mernst) [reply checker/src/main/java/org/checkerframework/checker/signedness/jdk.astub]: You could call clone on a String, or an Integer, or other types that are not arrays.

Comment (gokaco) [reply checker/src/main/java/org/checkerframework/checker/signedness/jdk.astub]: In order to expose the clone() method to clients, we have to declare it in our subclass with a more liberal access modifier, either public or package private. However, we cannot extend java.lang.Integer because it is a final class. Also note that java.lang.Integer does not implement Cloneable then how can we use clone() method with Integer or string etc?

Comment (mernst) [reply checker/src/main/java/org/checkerframework/checker/signedness/jdk.astub]: Yes, you have identified a serious and interesting limitation.  We are not sure how to deal with it.
I'm not sure it is a problem in your case, though.  Why do you need to change the contract of clone?  Isn't the contract on the Object class sufficient for all subclasses?

File: checker/src/main/java/org/checkerframework/checker/signedness/jdk.astub [hunk 4]
```diff
@@ -94,6 +98,11 @@ class Long {
     public static @Unsigned long remainderUnsigned(@Unsigned long dividend, @Unsigned long divisor);
 }
 
+class Object {
+    protected @PolySigned Object clone() throws CloneNotSupportedException;
```
Comment (mernst) [inline checker/src/main/java/org/checkerframework/checker/signedness/jdk.astub]: You have used only one `@PolySigned`, and there is no occurrence in a binding site. 
You need to write `@PolySigned` on the receiver as well.

File: checker/src/main/java/org/checkerframework/checker/signedness/jdk.astub (hunk 1 above)
Comment (gokaco) [reply checker/src/main/java/org/checkerframework/checker/signedness/jdk.astub]: I am not able to apply clone in Integer or String.

File: checker/src/main/java/org/checkerframework/checker/signedness/jdk.astub (hunk 4 above)
Comment (gokaco) [reply checker/src/main/java/org/checkerframework/checker/signedness/jdk.astub]: Yes, you are right. On the other hand, if we have to use clone in Integer or String then we again have to define the function in the program. I think for the time being we should not annotate it.

File: checker/src/main/java/org/checkerframework/checker/signedness/jdk.astub (hunk 1 above)
Comment (mernst) [reply checker/src/main/java/org/checkerframework/checker/signedness/jdk.astub]: Why not?  Can you show test code and the results?

File: checker/src/main/java/org/checkerframework/checker/signedness/jdk.astub (hunk 4 above)
Comment (mernst) [reply checker/src/main/java/org/checkerframework/checker/signedness/jdk.astub]: > if we have to use clone in Integer or String then we again have to define the function in the program.

Can you explain why?

File: checker/src/main/java/org/checkerframework/checker/signedness/jdk.astub (hunk 1 above)
Comment (gokaco) [reply checker/src/main/java/org/checkerframework/checker/signedness/jdk.astub]: If we simply run in java-
import java.lang.*; 
class ex2 { 

	private static Integer de(Integer ci){

		Integer b=ci.clone();
		return b;
	}
}
then it will give error.-
gsoc/ex2.java:6: error: clone() has protected access in Object
		Integer b=ci.clone();
		            ^
gsoc/ex2.java:6: error: incompatible types: Object cannot be converted to Integer
		Integer b=ci.clone();
		                  ^
2 errors

File: checker/src/main/java/org/checkerframework/checker/signedness/jdk.astub (hunk 4 above)
Comment (gokaco) [reply checker/src/main/java/org/checkerframework/checker/signedness/jdk.astub]: Like this-
import java.lang.*;
class Integer implements Cloneable{
	public Object clone()throws CloneNotSupportedException{
	return super.clone();  
	}
}
class ex2 {

		void play(){

		try{
		Integer ci=null;
		Integer b=(Integer)ci.clone();
		}
		catch(CloneNotSupportedException c){}
	}
}

File: checker/src/main/java/org/checkerframework/checker/signedness/jdk.astub (hunk 1 above)
Comment (gokaco) [reply checker/src/main/java/org/checkerframework/checker/signedness/jdk.astub]: 2nd error can be corectted by-
 Integer b=(Integer)ci.clone();
but how the 1st one.

File: checker/src/main/java/org/checkerframework/checker/signedness/jdk.astub (hunk 4 above)
Comment (gokaco) [reply checker/src/main/java/org/checkerframework/checker/signedness/jdk.astub]: If we try to implement clone inn Integer than we can do by this.-
import java.lang.*;
class Integer implements Cloneable{
	public Object clone()throws CloneNotSupportedException{
	return super.clone();  
	}
}
class ex2 {

		void play(){

		try{
		Integer ci=null;
		Integer b=(Integer)ci.clone();
		}
		catch(CloneNotSupportedException c){}
	}
}

but here Integer is a class with a name "Integer". This Integer is not the same Integer which we want. I googled but I am unable to find anything on how to use clone with it.

Comment (mernst) [reply checker/src/main/java/org/checkerframework/checker/signedness/jdk.astub]: Now I understand that the issue is that Integer does not implement Cloneable, so myInteger.clone() could never work, regardless of annotations.  Thanks for the explanation.  Removing the annotation on Object.clone() seems fine, in this case.

**Comment by mernst** [comment]
Thanks for this contribution!
